### PR TITLE
storage: gracefully handle shard finalization in persist_sink

### LINKED ISSUE #21314: thread 'timely:work-0' panicked at 'persist_sink(u1552): invalid upper! Tried to append batch
### What version of Materialize are you using?

3afbf2025f

### What is the issue?

This panic happened again on a Nightly run:

```
testdrive-materialized-1     | thread 'timely:work-0' panicked at 'persist_sink(u1552): invalid upper! Tried to append batch (Antichain { elements: [1692641868001] } -> Antichain { elements: [1692641869001] }) but upper is Antichain { elements: [] }. This is not a problem, it just means someone else was faster than us. We will try again with a new batch description.', /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0d30d47c959e97dbf-1/materialize/tests/src/storage/src/render/persist_sink.rs:1247:25 
```

Supposedly was fixed in https://github.com/MaterializeInc/materialize/pull/20847 but note that in this case the Antichain is empty.

https://buildkite.com/materialize/nightlies/builds/3416#_

### PR DESCRIPTION
**Pull request by petrosagg**
### Motivation

Our current approach to dropping a source results in a race between shard finalization (which happens in the controller) and dataflow shutdown (which happens in clusterd). If a source is created and dropped fast enough -or the two commands get sufficiently delayed- then it's possible for clusterd to attempt to start writing to a shard that the controller has already finalized.

This resulted in a panic in clusterd instead of a graceful shutdown of the dataflow.

This PR implements a solution to this problem by restarting the ingestion dataflow whenever an upper mismatch is encountered in the storage persist sink implementation. This is a stopgap solution until concurrent ingestion is propertly implemented and such upper mismatches get handled inside the operators.

Fixes #21314

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by guswynn — COMMENTED** [review]
Won't this trigger this panic: https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/storage_state.rs#L724 ?

**Review by guswynn — COMMENTED** [review]
Won't this trigger this panic: https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/storage_state.rs#L724 ?

**Comment by petrosagg** [comment]
Hm, can you explain how that panic will be triggered?

**Comment by guswynn** [comment]
@petrosagg 

> If a source is created and dropped fast enough -or the two commands get sufficiently delayed-

I imagine that we could receive the delayed `AllowCompaction([])` command in storage_state right before we send the `SuspendAndRestart` command into the command stream from this operator, we will hit this bug

Actually, there may be a more general bug here, where _any_ `SuspendAndRestart` can interleave with a drop command?

**Comment by petrosagg** [comment]
> Actually, there may be a more general bug here, where any `SuspendAndRestart` can interleave with a drop command?

I reached the same conclusion, yes. I need to think about this

**Comment by guswynn** [comment]
@petrosagg resolving this also required to fix one of the sink panics we are periodically seeing, as well, i believe

**Comment by petrosagg** [comment]
@guswynn I removed the assertion as we discussed as there is nothing better we can do at the moment

File: src/storage/src/storage_state.rs [hunk 1]
```diff
@@ -711,23 +711,7 @@ impl<'w, A: Allocate> Worker<'w, A> {
...
-                if self
-                    .storage_state
-                    .ingestions
-                    .values()
-                    .any(|v| v.source_exports.contains_key(&id))
-                {
-                    return;
                 }
-
-                panic!("got InternalStorageCommand::SuspendAndRestart for something that is not a source or sink: {id}");
```
Comment (guswynn) [inline src/storage/src/storage_state.rs:730]: if you want, we can leave an `error!` or `warn!` logline and get a sentry event to track this issue

**Review by guswynn — APPROVED** [review]
nice

File: src/storage/src/storage_state.rs (hunk 1 above)
Comment (petrosagg) [reply src/storage/src/storage_state.rs]: Good idea, added!

**Pull request merged**